nexus/src/app/instance_platform.rs

```diff
@@ -0,0 +1,481 @@
+            if map
+                .insert(
+                    disk.id(),
+                    CrucibleDisk { device_name, device, backend },
+                )
+                .is_some()
+            {
+                return Err(Error::internal_error(&format!(
+                    "instance has multiple attached disks with ID {}",
+                    disk.id()
+                )));
+            }
+        }
+
+        Ok(Self(map))
+    }
+}
+
+/// A list of named components to add to an instance's spec.
+struct Components(HashMap<String, ComponentV0>);
+
+impl Default for Components {
+    /// Adds the default set of VM components that are expected to exist for all
+    /// Oxide VMs (serial ports attached to COM1-COM4 and a paravirtualized
+    /// panic device).
+    fn default() -> Self {
+        let map = [
+            (
+                component_names::COM1,
+                ComponentV0::SerialPort(SerialPort {
+                    num: SerialPortNumber::Com1,
+                }),
+            ),
+            (
+                component_names::COM2,
+                ComponentV0::SerialPort(SerialPort {
+                    num: SerialPortNumber::Com2,
+                }),
+            ),
+            (
+                component_names::COM3,
+                ComponentV0::SerialPort(SerialPort {
+                    num: SerialPortNumber::Com3,
+                }),
+            ),
+            (
+                component_names::COM4,
+                ComponentV0::SerialPort(SerialPort {
+                    num: SerialPortNumber::Com4,
+                }),
+            ),
+            (
+                component_names::PVPANIC,
+                ComponentV0::QemuPvpanic(QemuPvpanic { enable_isa: true }),
+            ),
+        ]
+        .into_iter()
+        .map(|(k, v)| (k.to_owned(), v))
+        .collect();
+
+        Self(map)
+    }
+}
+
+impl Components {
+    /// Adds a named component to this component list. Returns a 500 error if
+    /// the component name was already present in the list.
+    fn add(
+        &mut self,
+        key: String,
+        component: ComponentV0,
+    ) -> Result<(), Error> {
+        use std::collections::hash_map::Entry;
+
+        // Component names are an implementation detail internal to this module,
+        // so they should always be unique.
+        match self.0.entry(key) {
+            Entry::Occupied(occupied_entry) => {
+                Err(Error::internal_error(&format!(
+                    "duplicate instance spec component key: {}",
+                    occupied_entry.key()
+                )))
+            }
+            Entry::Vacant(vacant_entry) => {
+                vacant_entry.insert(component);
+                Ok(())
+            }
+        }
+    }
+
+    /// Adds the set of disks in the supplied disk list to this component
+    /// list.
+    fn add_disks(&mut self, disks: DisksById) -> Result<(), Error> {
```

Comment:
take it or leave it: perhaps we ought to stick a
```suggestion
    fn add_disks(&mut self, disks: DisksById) -> Result<(), Error> {
        self.0.reserve(disks.0.len() * 2);
```
since we know how many entries we're going to add up front?